flytekit/extras/tensorflow/records.py

```diff
@@ -0,0 +1,133 @@
+from dataclasses_json import dataclass_json
+from tensorflow.core.example import example_pb2
+from tensorflow.python.data.ops.readers import TFRecordDatasetV2
+from typing_extensions import Annotated, get_args, get_origin
+
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+from flytekit.types.directory import TFRecordsDirectory
+from flytekit.types.file import TFRecordFile
+
+T = TypeVar("T")
+
+
+@dataclass_json
+@dataclass
+class TFRecordDatasetConfig:
+    """
+    TFRecordDatasetConfig is the config used during the creating tf.data.TFRecordDataset comprising
+    record of one or more TFRecord files.
+
+    Args:
+      compression_type:A scalar evaluating to one of "" (no compression), "ZLIB", or "GZIP".
+      buffer_size:the number of bytes in the read buffer.If None, a sensible default for both local and remote file systems is used.
+      num_parallel_reads:the number of files to read in parallel. If greater than one, the records of files read in parallel are outputted in an interleaved order.
+      name:A name for the operation.
+    """
+
+    compression_type: Optional[str] = None
+    buffer_size: Optional[int] = None
+    num_parallel_reads: Optional[int] = None
+    name: Optional[str] = None
+
+
+def extract_metadata(t: Type[T]) -> Tuple[T, TFRecordDatasetConfig]:
+    metadata = TFRecordDatasetConfig()
+    if get_origin(t) is Annotated:
+        base_type, metadata = get_args(t)
+        if isinstance(metadata, TFRecordDatasetConfig):
+            return base_type, metadata
+        else:
+            raise TypeTransformerFailedError(f"{t}'s metadata needs to be of type Dict")
+    return t, metadata
+
+
+class TensorflowRecordsTransformerBase(TypeTransformer, Generic[T]):
+    """
+    TypeTransformer that supports serialising and deserialising to and from TFRecord file.
+    https://www.tensorflow.org/tutorials/load_data/tfrecord
+    """
+
+    def get_literal_type(self, t: Type[T]) -> LiteralType:
+        return LiteralType(
+            blob=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE,
+            )
+        )
+
+    def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], expected: LiteralType) -> Literal:
+        meta = BlobMetadata(
+            type=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE,
+            )
+        )
+        local_dir = ctx.file_access.get_random_local_directory()
+        remote_path = ctx.file_access.get_random_remote_directory()
+        if isinstance(python_val, TFRecordsDirectory):
```

Comment:
Shouldn't the `python_val` be `tf.train.Example`?

Reply:
I think the `TypeTransformer` has to be such that it subclasses both `TFRecordFile` and `TFRecordsDirectory`. In the `to_literal` method, it needs to validate whether the type of `python_val` is `tf.train.Example` or a list of `tf.train.Example`s and thereby, return a blob with [dimensionality set to single or multipart](https://github.com/flyteorg/flytekit/blob/9e4069058845f018f410703ec7c4e3971a9ee85a/flytekit/models/core/types.py#L37-L39).